### PR TITLE
flower_reader.py

### PR DESCRIPTION
**Pull request by amymhaddad**
Instructions: Create a function that opens the flowers.txt, reads every line in it, and saves it as a dictionary. The main (separate) function should take user input (user's first name and last name) and parse the user input to identify the first letter of the first name. It should then use it to print the flower name with the same first letter (from dictionary created in the first function).

Ex: >>> Enter your First [space] Last name only: Bill Newman
>>> Unique flower name with the first letter: Bellflower

### CODE REVIEW

File: flower_reader.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Thu May 10 15:34:22 2018
+
+@author: amyhaddad
+"""
+
+FILENAME = 'flowers.txt'
```
Comment (amymhaddad) [inline flower_reader.py:12]: @paulghaddad I need to break down this file into a main file and an exercise file. But before I'd like this first iteration to just look at the code and see if I need to refactor further. 

Also, thoughts on the constants? Pylint suggested I change them to constants.

Comment (paulghaddad) [reply flower_reader.py:12]: @amymhaddad I agree on this being a constant. In the real world, you'd probably make this take an argument from the command line:

```
python flower_reader.py flowers.txt
```

But that's for another day.

File: flower_reader.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+USER_NAME = input('Enter your first and last name: ')
```
Comment (paulghaddad) [inline flower_reader.py]: I'm not entirely sure this needs to be a constant unless Pylint says so. I'd personally use a variable here.

File: flower_reader.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+FLOWERS = {}
+FIND_FIRST_LETTER_OF_USER_NAME = lambda USER_NAME: USER_NAME[0].title()
```
Comment (paulghaddad) [inline flower_reader.py]: And here. Although nice lambda!

File: flower_reader.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    """Read in a file and create a dictionary"""
+
+    list_of_flowers = []
+    with open(FILENAME) as f_object:
+        contents = f_object.readlines()
+        for flower in contents:
+            split_flower_content = flower.split(" ", 1)
+            list_of_flowers.append(split_flower_content)
+
+        for flower_types in list_of_flowers:
```
Comment (paulghaddad) [inline flower_reader.py]: Maybe `flower_types` is better named as `flower`?

File: flower_reader.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            letter = flower_types[0]
+            flower_name = flower_types[1]
+            FLOWERS[letter] = flower_name
+
+        return FLOWERS
+
+def match_letter_to_flower(USER_NAME, FILENAME):
+    """Match the first letter of user's name to the flower type"""
+
+    flower_dictionary = create_flower_dictionary(FILENAME)
```
Comment (paulghaddad) [inline flower_reader.py]: Nice extracted function!

File: flower_reader.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    first_letter = FIND_FIRST_LETTER_OF_USER_NAME(USER_NAME)
+
+    for letters, flower_names in flower_dictionary.items():
```
Comment (paulghaddad) [inline flower_reader.py:37]: Really nice loop over the dictionary (with tuple unpacking)! 👏

File: flower_reader.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+@author: amyhaddad
+"""
+
+FILENAME = 'flowers.txt'
+USER_NAME = input('Enter your first and last name: ')
+FLOWERS = {}
+FIND_FIRST_LETTER_OF_USER_NAME = lambda USER_NAME: USER_NAME[0].title()
+
+def create_flower_dictionary(FILENAME):
```
Comment (paulghaddad) [inline flower_reader.py]: use `filename` as the parameter name, remember this is _not_ the actual item you'll pass in when calling the function--that's the _argument_.

File: flower_reader.py (hunk 4 above)
Comment (paulghaddad) [inline flower_reader.py]: So the only issue with your algorithm is for each line, you're appending the flower content to `list_of_flowers`. Then on the loop starting on line 24, you're iterating through this list and modifying your dictionary. But notice that `list_of_flowers` will just get bigger and bigger and you'll redo work in building the dictionary. You don't need to store this list outside the main with on line 18. Just do this work in your existing loop on line 20.

**Comment by amymhaddad** [comment]
@paulghaddad I updated this exercise by condensing the first function (removing the 2nd for loop)

File: flower_reader.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+"""
+Created on Thu May 10 15:34:22 2018
+
+@author: amyhaddad
+"""
+
+FILENAME = 'flowers.txt'
+FLOWERS = {}
+user_name = input('Enter your first and last name: ')
+find_first_letters_of_user_name = lambda user_name: user_name[0].title()
```
Comment (paulghaddad) [inline flower_reader.py:14]: good lambda! 💯

File: flower_reader.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Thu May 10 15:34:22 2018
+
+@author: amyhaddad
+"""
+
+FILENAME = 'flowers.txt'
+FLOWERS = {}
```
Comment (paulghaddad) [inline flower_reader.py]: This is a good constant, similar to what we learned in Ned's talk.

**Review by paulghaddad — CHANGES_REQUESTED** [review]
This is great work. A few comments, but this should be a pretty quick update.

**Comment by amymhaddad** [comment]
@paulghaddad Updated the parameter with the more general "file_name". Ready for final review.

File: flower_reader.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+def create_flower_dictionary(file_name):
+    """Read in a file and create a dictionary"""
+
+    flowers = {}
+
+    with open(file_name) as f_object:
+        contents = f_object.readlines()
+        for flower in contents:
+            split_flower_content = flower.split(" ", 1)
+            letter = split_flower_content[0]
```
Comment (paulghaddad) [inline flower_reader.py:25]: From the looks of this code, it appears to split on the space and the first split would then contain the colon, such as `A:` and the second the flower name: `African Daisy`. Thus, I'm thinking you want to split with `flower.split(": ", 1)`. But I might be wrong. And if this is the case, you wouldn't have to use the `maxsplit` argument of `1`. (although I learned something there!)

**Review by paulghaddad — CHANGES_REQUESTED** [review]
Excellent! Just one minor tweak!

File: flower_reader.py (hunk 10 above)
Comment (amymhaddad) [reply flower_reader.py:25]: @paulghaddad It does not appear that changing the split from ":" to " " affects anything. So I just kept it as it was. Does that work?

Comment (paulghaddad) [reply flower_reader.py:25]: 👍

**Review by paulghaddad — APPROVED** [review]
🎉

**Pull request merged**